network/blob_service.go

```diff
@@ -0,0 +1,180 @@
+package network
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/ipfs/go-bitswap"
+	bsnet "github.com/ipfs/go-bitswap/network"
+	blocks "github.com/ipfs/go-block-format"
+	blockservice "github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	provider "github.com/ipfs/go-ipfs-provider"
+	"github.com/ipfs/go-ipfs-provider/simple"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-libp2p-core/routing"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+)
+
+type Blob = blocks.Block
+
+var NewBlob = blocks.NewBlock
+
+// BlobGetter is the common interface shared between blobservice sessions and
+// the blobservice.
+type BlobGetter interface {
+	// GetBlob gets the requested blob.
+	GetBlob(ctx context.Context, c cid.Cid) (Blob, error)
+
+	// GetBlobs does a batch request for the given cids, returning blobs as
+	// they are found, in no particular order.
+	//
+	// It may not be able to find all requested blobs (or the context may
+	// be canceled). In that case, it will close the channel early. It is up
```

Comment:
⚠️ In the above cases, except case 1, all the other cases is caused by certain error, but the API doesn't have a way to communicate about this error, it relies on returning partial data to give the caller clue that something wrong happened.  And there is no way to explicitly communicate what was the exact error. 

This is unlike the `GetBlob` API where it will return an error about why the blob for the given ID can't be retrieved. 

I understand returning a `chan Blob` is probably good for streaming process, but I found missing the error reporting is a bit concerning to me. 

I had another idea for you to consider, this might allow both streaming process and error reporting:
Instead of letting `GetBlobs` return a `<- chan Blob`, we could let it take a stream processing function `func (blob, error) error`, so that when my function gets called, you will either give me a blob or an error.

Reply:
> I had another idea for you to consider, this might allow both streaming process and error reporting:
Instead of letting GetBlobs return a <- chan Blob, we could let it take a stream processing function func (blob, error) error, so that when my function gets called, you will either give me a blob or an error.

This is actually similar to what I originally started with, but after working with that for a while, I decided that returning a channel has much more flexibility and was actually easier to work with.

case 3 shouldn't really be an issue, the code properly handles this case and anyways it should never occur.

case 2 could in fact *only* happen when the context is canceled (or times out). This is not apparent from the API, but is true based on the underlying implementation.

case 4 could never happen.

case 5 is just a special case of case 2.

Reply:
@Ullaakut For example, you can cancel iteration over the channel as soon as you want. If we build another layer of abstraction on top of this, then we can only cancel the context, but we cannot signal the underlying implementation to stop consuming the rest of the blobs that are already in the channel.

There are other reasons, for example it is not as easy to keep track of the order in which you receive items, because from the perspective of the callback you don't know how many blobs have already come before you.

Initially the callback seems like a good idea, and that's why I initially implemented it that way. However, when you work with it more, you may realize that it sacrifice a lot of flexibility.